staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -1231,11 +1313,32 @@ func (c *cacheWatcher) add(event *watchCacheEvent, timer *time.Timer) bool {
 
 	closeFunc := func() {
 		// This means that we couldn't send event to that watcher.
-		// Since we don't want to block on it infinitely,
-		// we simply terminate it.
+		// Since we don't want to block on it infinitely, we simply terminate it.
+
+		// If the bookmarkAfterResourceVersionReceived hasn't been received,
+		// we can safely terminate the watcher. Because the client is waiting
+		// for this specific bookmark, and we even haven't received one.
+		if !c.wasBookmarkAfterResourceVersionReceived() {
+			klog.V(1).Infof("Forcing %v watcher close due to unresponsiveness: %v. len(c.input) = %v, len(c.result) = %v", c.objectType.String(), c.identifier, len(c.input), len(c.result))
+			metrics.TerminatedWatchersCounter.WithLabelValues(c.objectType.String()).Inc()
+			c.forget(false)
+			return
+		}
+		// We have seen a bookmark, but it hasn't been sent to the client yet.
+		// That means we should drain the input buffer which contains
+		// the bookmarkAfterResourceVersion we want. We do that to make progress
+		// as clients can re-establish a new watch with the given RV and receive
+		// further notifications.
+		if !c.wasBookmarkAfterResourceVersionSentLocked() {
+			c.forget(true)
+			return
```

Comment:
Same about the log - we should also write it in all cases, just add an additional info whether this deletion is graceful or not.

So I would do something like (the more I'm thinking about it, the more I think that the state-machine idea is a good one).
```
// If we didn't even receive the bookmark, we can safely terminate the watcher.
// If we received the bookmark but didn't yet send it, we should try to send it by shutting
// down gracefully to allow the client to make the progress.
// If we already send a bookmark, we can again safely terminate the watcher.
graceful := c.state == BOOKMARK_RECEIVED
klog.V(1).Infof("Forcing ... graceful=%t ...)
c.foget(graceful)
```

Reply:
With your current changes from the last commit - this would now simplify the code here.